src/api/app/policies/application_policy.rb

```diff
@@ -1,7 +1,10 @@
 class ApplicationPolicy
   attr_reader :user, :record
 
+  PLEASE_LOGIN = 'Please login to access the requested page.'.freeze
+
   def initialize(user, record, opts = {})
+    ensure_logged_in!(user, opts)
```

Comment:
The idea is to extend the Base class to deal with anonymous user in a backward compatible way.
Later on we could remove the check below (after making sure that the API works well) and further refactor the code in order to default `:ensure_logged_in ` to `true`.

Reply:
I really liked this approach, thank you!

Reply:
I wonder if this is really needed since in the following line, we're already handling anonymous users or not, if we pass the option `user_optional`. It's not checking for `user.is_nobody?` though. This could be changed in the following line.

Was `ensure_logged_in!` introduced to prevent potential errors with the `reason` option for `Pundit::NotAuthorizedError`? Is this what you meant with backward compatible?

Reply:
Changing the line below and relying on `:user_optional` flag has an impact on all the already existing Policies which relies on it (pretty much in the API), since I wanted to reduce the scope of the code to check in this PR I've opted for introducing a new flag.
The idea is that eventually we'll migrate all the codebase to the new flag, but we can do this more gradually.

Moreover we introduced the `reason` attribute and we should (probably) also extend the API `rescue_handler` to deal with it.

Reply:
The `reason` attribute on `Pundit::NotAuthorizedError` is a purely additive one. Everything will keep behaving like it did before, even if the `reason` attribute is set. To stay backward compatible, we can set the `reason` attribute while still setting an exception message. It could be done like this, on the line 8 of this file (so directly below this comment):
```
    raise Pundit::NotAuthorizedError(reason: ANONYMOUS_USER), 'must be logged in' unless user || opts[:user_optional]
```

It's not needed though since nothing relies on the exception's message, so I would just go for this:
```
    raise Pundit::NotAuthorizedError(reason: ANONYMOUS_USER) unless user || opts[:user_optional]
```

`RescueHandler` (for the API) will simply ignore the `reason` attribute on `Pundit::NotAuthorizedError`. It's relying on the `record` and `query` (if present). See below: https://github.com/openSUSE/open-build-service/blob/bbdaa4907875ea3fb76c3813b15e2e72a6ddb945/src/api/app/controllers/concerns/rescue_handler.rb#L46-L68

`Webui::RescueHandler` will handle the `reason` attribute whenever it's provided with the changes introduced by this PR in `Webui::RescueHandler#unauthorized_message`. This is the right approach.

Beside those rescue handlers, there are two controllers which also rescue from `Pundit::NotAuthorizedError` and they are both not relying on the exception message. Why are they not using the rescue handlers? This is a question for another time...

So to conclude, this is why I would be against introducing another method to handle anonymous users.

Reply:
@dmarcoux I agree with you that `reason` is not an issue 👍 .

My main concern was that if we change the `:user_optional` check in order to deal with anonymous user the existing logic changes for all the `ApplicationPolicy` subclasses, in particular all the Policies which have `:user_optional` not set (which is the default) will start raising a `NotAuthorizedError` when `user.is_nobody?` and that's what I meant with *backward compatibility*.

Since I was there I took a look and tried to figure out if the above was an issue therefore I've checked all the actions which do both `skip_before_action :require_login` and call `authorize` in order to figure out if any of them wouldn't work with the new change: looks like all of them will work well with this change therefore I've just pushed your proposal.
I hope I was clear this time and sorry for omitting some details.

*Side note*: some specs are failing now but It's related to the fact that the anonymous check is in the `initializer`, if you approve the changes I'll fix them 😄

Reply:
@dmarcoux I've edited the message above.

Reply:
We have two approaches. The first one, we go with the backward compatible way and later deal with it. This limits the scope of this PR. The second way is dealing with all policies requiring changing due to the behavior of the `ApplicationPolicy`. This increases the scope of this PR a lot. I underestimated the increase of this approach.

Let's go with your first proposal. Sorry for the back and forth, I thought it wouldn't be so problematic to go with the second approach. I was wrong.

Reply:
@dmarcoux No worries at all! I also didn't like the idea of adding another flag but reducing the scope of the PR is a good trade-off.
Was good that you pointed out the issue so that we had the opportunity to dig deeper into that option as well 👍 .

I've just rolled back the last changes.